certora/specs/GuardianStorage.spec

```diff
@@ -0,0 +1,378 @@
+        }
+    }
+
+rule storeHookPreservesInvariants(address wallet, address key, address value) {
+    // These are checked in the hook.
+    require key != NULL;
+    require reach(wallet, value, key) => value == SENTINEL; //, "list is cyclic";
+
+    // Invariants that hold even in the middle
+    requireInvariant reach_null();
+    requireInvariant reach_invariant();
+
+    address someKey;
+    address someWallet;
+    require reach_succ(someWallet, someKey, ghostGuardians[someWallet][someKey]);
+    require ghostSuccCount(someWallet, someKey) == count_expected(someWallet, someKey);
+    ghostGuardians[wallet][key] = value;
+    havoc reach assuming updateSucc(wallet, key, next_or_null(value));
+    mathint countDiff = count_expected(wallet, key) - ghostSuccCount(wallet, key);
+    havoc ghostSuccCount assuming updateGhostSuccCount(wallet, key, countDiff);
+    assert reach_succ(someWallet, someKey, ghostGuardians[someWallet][someKey]), "reach_succ violated after guardians update";
+    assert ghostSuccCount(someWallet, someKey) == count_expected(someWallet, someKey);
+
+    // assert also the invariants used above
+    assert (forall address W. forall address X. reach(W, X, NULL));
+    assert (forall address W. forall address X. forall address Y. forall address Z.
+        reach(W, X, X)
+        && (reach(W, X, Y) && reach(W, Y, X) => X == Y)
+        && (reach(W, X, Y) && reach(W, Y, Z) => reach(W, X, Z))
+        && (reach(W, X, Y) && reach(W, X, Z) => (reach(W, Y, Z) || reach(W, Z, Y)))
+    );
+}
+
+rule isGuardianDoesNotRevert {
+    address addr;
+    isGuardian@withrevert(safeContract, addr);
+    assert !lastReverted, "isGuardian should not revert";
+}
+
+rule isGuardianNotSentinal {
+    address addr;
+    require addr == SENTINEL;
+    bool result = isGuardian(safeContract, addr);
+    assert result == false, "SENTINEL must not be guardian";
+}
+
+rule isGuardianInList {
+    address addr;
+    env e;
+    require safeContract.isModuleEnabled(e.msg.sender);
+    require addr != SENTINEL;
+    bool result = isGuardian(safeContract, addr);
+    assert result == (ghostGuardians[safeContract][addr] != NULL), "isGuardian returns wrong result";
+}
+
+rule addGuardianChangesEntries {
+    address other;
+    address toAdd;
+    uint256 threshold;
+    env e;
+    require safeContract.isModuleEnabled(e.msg.sender);
+
+    requireInvariant reach_null();
+    requireInvariant reach_invariant();
+    requireInvariant inListReachable();
+    requireInvariant reachableInList();
+    requireInvariant countZeroIffListEmpty();
+    require other != toAdd;
+    bool isGuardianOtherBefore = isGuardian(safeContract, other);
+    addGuardianWithThreshold(e, safeContract, toAdd, threshold);
+
+    assert isGuardian(safeContract, toAdd), "addGuardian should add the given guardian";
+    assert isGuardian(safeContract, other) == isGuardianOtherBefore, "addGuardian should not remove or add other guardians";
+    satisfy true;
```

Comment:
what does this line do?

Reply:
good catch. I had added it initially when we started with FV. Removed it now. https://github.com/5afe/CandideWalletContracts/pull/20/commits/584ff6545591c33f71a1d6b1db136f70824e6337